pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go

```diff
@@ -0,0 +1,282 @@
+				d.log.Debug("Unknown angular pattern", "name", pattern.Name, "type", pattern.Type, "error", err)
+				continue
+			}
+			// Other error, do not ignore it
+			finalErr = errors.Join(finalErr, err)
+		}
+		detectors = append(detectors, ad)
+	}
+	if finalErr != nil {
+		return nil, finalErr
+	}
+	return detectors, nil
+}
+
+// fetch fetches the angular patterns from GCOM and returns them as GCOMPatterns.
+// Call detectors() on the returned value to get the corresponding detectors.
+func (d *Dynamic) fetch(ctx context.Context) (GCOMPatterns, error) {
+	st := time.Now()
+
+	reqURL, err := url.JoinPath(d.baseURL, gcomAngularPatternsPath)
+	if err != nil {
+		return nil, fmt.Errorf("url joinpath: %w", err)
+	}
+
+	d.log.Debug("Fetching dynamic angular detection patterns", "url", reqURL)
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, reqURL, nil)
+	if err != nil {
+		return nil, fmt.Errorf("new request with context: %w", err)
+	}
+	resp, err := d.httpClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("http do: %w", err)
+	}
+	defer func() {
+		if closeErr := resp.Body.Close(); closeErr != nil {
+			d.log.Error("Response body close error", "error", err)
+		}
+	}()
+	var out GCOMPatterns
+	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
+		return nil, fmt.Errorf("json decode: %w", err)
+	}
+	d.log.Debug("Fetched dynamic angular detection patterns", "patterns", len(out), "duration", time.Since(st))
+	return out, nil
+}
+
+// setDetectors sets the detectors by acquiring the lock first.
+func (d *Dynamic) setDetectors(newDetectors []angulardetector.AngularDetector) {
+	d.mux.Lock()
+	d.detectors = newDetectors
+	d.mux.Unlock()
+}
+
+// updateDetectors fetches the patterns from GCOM, converts them to detectors,
+// stores the patterns in the database and update the cached detectors.
+func (d *Dynamic) updateDetectors(ctx context.Context) error {
+	// Fetch patterns from GCOM
+	patterns, err := d.fetch(ctx)
+	if err != nil {
+		return fmt.Errorf("fetch: %w", err)
+	}
+
+	// Convert the patterns to detectors
+	newDetectors, err := d.patternsToDetectors(patterns)
+	if err != nil {
+		return fmt.Errorf("patterns convert to detectors: %w", err)
+	}
+
+	// Update store only if the patterns can be converted to detectors
+	if err := d.store.Set(ctx, patterns); err != nil {
+		return fmt.Errorf("store set: %w", err)
+	}
+
+	// Update cached detectors
+	d.setDetectors(newDetectors)
+	return nil
+}
+
+// setDetectorsFromCache sets the in-memory detectors from the patterns in the store.
+func (d *Dynamic) setDetectorsFromCache(ctx context.Context) error {
+	var cachedPatterns GCOMPatterns
+	rawCached, ok, err := d.store.Get(ctx)
+	if !ok {
+		// No cached value found, do not alter in-memory detectors
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("get cached value: %w", err)
+	}
+	// Try to unmarshal, convert to detectors and set local cache
+	if err := json.Unmarshal([]byte(rawCached), &cachedPatterns); err != nil {
```

Comment:
Should this logic perhaps live in the store to provide a nicer API and make the store operate on and return strongly typed types rather than `any`?